[Serialization] Serialize archetypes as generic environment + interface type

DougGregor opened the pull request:
Rather than serializing the complete structure of all archetypes
(which is completely redundant), serialize a reference to their owning
generic environment as well as their interface type. The archetype
itself will be reconsituted by mapping the interface type into that
generic environment.

slavapestov commented:
Wow, nice!

slavapestov commented on lib/AST/GenericEnvironment.cpp:
```diff
@@ -40,24 +40,44 @@ GenericEnvironment::GenericEnvironment(
+static unsigned declContextDepth(const DeclContext *dc) {
+  unsigned depth = 0;
+  while (auto parentDC = dc->getParent()) {
+    ++depth;
```
This is wrong for extensions of nested generic types. If you hit an extension dc you have to look at it's generic parameter list, and add up the number of outer parameter lists.

DougGregor replied:
Good catch, thanks!

slavapestov commented on include/swift/AST/Decl.h:
```diff
@@ -4086,13 +4089,17 @@ class VarDecl : public AbstractStorageDecl {
-    return !!typeInContext;
+    // We have a type if either the type has been computed already or if
+    // this is a deserialized declaration with an interface type.
+    return typeInContext ||
```
I wanted to refactor things so that we never call getType() on deserialized ParamDecls, or VarDecls not in local context. Since VarDecls in local context that are not ParamDecls do not get seralized, this means we should be able to kill this hack altogether.

slavapestov commented:
@DougGregor do we need to serialize the types of Patterns at all? Do we ever look at PatternBindingDecls once we've built up our VarDecls, accessors and nominal type initializers?

DougGregor committed:
[Serialization] Serialize archetypes as generic environment + interface type.

Rather than serializing the complete structure of all archetypes
(which is completely redundant), serialize a reference to their owning
generic environment as well as their interface type. The archetype
itself will be reconsituted by mapping the interface type into that
generic environment.

DougGregor commented:
@huonw I *think* this will start working when https://github.com/apple/swift/pull/7313 lands.

jrose-apple commented:
To answer Slava's question from months ago, I'm 95% sure we do not actually look at the PBDs, and do not need to serialize them at all. That's a change we could consider making now, "early" in Swift 4.

DougGregor commented:
Updated title and description to reflect the current state of affairs. Most of what was in the original PR landed separately, so it's down to this last bit to actually change the serialized representation of `ArchetypeType`. It's running into some trouble that will be helped by https://github.com/apple/swift/pull/7313, but I have some debugging to do as well.

DougGregor committed:
[Serialization] Don't cross-reference module-scope generic parameter decls.

With serialization of the interface types of archetypes, we can end up
cross-referencing module-scope generic parameter declarations when
writing those interface types. Desugar such types on serialization to
fix a regression introduced by the change to generic environment
serialization.

DougGregor committed:
[Type checker] Keep a type checker alive during name binding.

Name lookup can cause import of some Clang declaration, which in turn
need a type checker. Make sure a type checker exists at this time.

This is a baby step toward having the type checker always available,
which is really the end goal. For now, we need this for better generic
environment (de-)serialization.

DougGregor committed:
[Clang importer] Compute the requirement signature of imported protocols.

We can also avoid walking the members of said protocols in the
archetype builder and when looking for nested types, because of course
there are no associated types in imported Objective-C protocols.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please test and merge

DougGregor commented:
@swift-ci please test and merge

DougGregor commented:
@swift-ci please smoke test OS X

DougGregor commented:
@swift-ci please smoke test OS X
